zilliqa/src/consensus.rs

```diff
@@ -999,83 +1031,148 @@ impl Consensus {
         Ok(None)
     }
 
-    fn propose_new_block_at(
+    /// Finalise the early Proposal.
+    /// This should only run after majority QC is available.
+    /// It applies the rewards and produces the final Proposal.
+    fn finish_early_proposal_at(
         &self,
         state: &mut State,
-        transaction_pool: &mut TransactionPool,
-        votes: &mut BTreeMap<Hash, BlockVotes>,
-    ) -> Result<Option<(Block, Vec<VerifiedTransaction>)>> {
-        let num = self
-            .block_store
-            .get_highest_block_number()
-            .unwrap()
-            .unwrap();
-        let block = self.get_block_by_number(num).unwrap().unwrap();
+        votes: &BTreeMap<Hash, BlockVotes>,
+        proposal: Block,
+    ) -> Result<Option<Block>> {
+        // Retrieve parent block data
+        let parent_block = self
+            .get_block(&proposal.parent_hash())?
+            .context("missing parent block")?;
+        let parent_block_hash = parent_block.hash();
+        let parent_block_view = parent_block.view();
+
+        // Should have supermajority by now.
+        // Retrieve the committee - for rewards
+        let committee = state.get_stakers_at_block_raw(&parent_block)?;
+        let (signatures, cosigned, _, _) = votes
+            .get(&parent_block_hash)
+            .context("tried to finalise a proposal without any votes")?;
+
+        // Compute the majority QC
+        let qc = self.qc_from_bits(parent_block_hash, signatures, *cosigned, parent_block_view);
+
```

Comment:
At the moment, the dummy QC is only created once, at the earliest Vote that has a populated mempool, not each time a Vote is received. But yes, if we ever change that code so that we continuously populate the Proposal when every Vote is received, then it makes sense not to do it.

Anyway, I'll try to replace the dummy QC with a proper dummy that doesn't compute the AggSig e.g. it has an Identity signature.

Reply:
you're right, it's created only once, my bad